### PR TITLE
Add Azure Linux updating pipeline

### PR DESCRIPTION
**Pull request by mertakman**
This PR adds a new azurelinux pipeline to the go-infra repository, enhancing CI/CD for Azure Linux environments.

Changes
- Added: new update-azure-linux pipeline for automatically opening a pull request on Azure Linux for updating spec files

Test Pipeline Run 
https://dev.azure.com/dnceng/internal/_build/results?buildId=2525925&view=results

### CODE REVIEW

File: eng/pipelines/update-azure-linux-pipeline.yml
Comment (gdams) [inline eng/pipelines/update-azure-linux-pipeline.yml:81]: this is currently failing, I think we need to provide some more parameters such as the pipeline name etc

Comment (dagood) [reply eng/pipelines/update-azure-linux-pipeline.yml:81]: Yeah, it should look like this:

https://github.com/microsoft/go-infra/blob/f0ee8dc8eb530988cbb8a971ce9290b8a9484704/eng/pipelines/steps/release-build-steps.yml#L155-L165

(Although here, it can use a parameter rather than variable for the run ID (build ID).)

To test it out, copy the number from the `buildId=###` part of the URL of a microsoft-go build.

File: eng/pipelines/update-azure-linux-pipeline.yml
Comment (dagood) [inline eng/pipelines/update-azure-linux-pipeline.yml:10]: I don't think approveAheadOfTime is needed for this pipeline. `release-go-images-pipeline.yml` has an approval step because once that pipeline runs, it directly changes the production images in MAR. The approval means the release owner/dev has a chance to check things over first. (Specifically: did innerloop tests pass?)

This pipeline submits a PR, but it doesn't directly change any production assets. The change has to go through PR, and the release owner/dev can close the PR (or simply not merge it) if there's a problem that comes up.

I think deleting this parameter, everything in `- ${{ if eq(parameters.approveAheadOfTime, false) }}:`, and `retryInstructionsFlags` is fine, to simplify the pipeline.

File: eng/pipelines/update-azure-linux-pipeline.yml
Comment (dagood) [inline eng/pipelines/update-azure-linux-pipeline.yml:88]: Update `displayName`?

File: eng/pipelines/update-azure-linux-pipeline.yml
Comment (dagood) [inline eng/pipelines/update-azure-linux-pipeline.yml:87]: How does this find the assets.json file?

File: eng/pipelines/update-azure-linux-pipeline.yml
Comment (dagood) [inline eng/pipelines/update-azure-linux-pipeline.yml:35]: What's this variable used for?

There's already a variable in the releasego.yml template that I would expect to serve this purpose:

https://github.com/microsoft/go-infra/blob/f0ee8dc8eb530988cbb8a971ce9290b8a9484704/eng/pipelines/jobs/releasego.yml#L37-L40

**Review by dagood — COMMENTED** [review]
The comment about `- download: build` also needs to be addressed.

It would also be nice to have a link to a working test build in the PR description so we can take a look. I try to always include one (or two if I have to).

File: eng/pipelines/update-azure-linux-pipeline.yml
Comment (gdams) [reply eng/pipelines/update-azure-linux-pipeline.yml:35]: I think Mert added it for testing, looks like we can remove it and rely on `buildAssetJsonFile`

File: eng/pipelines/update-azure-linux-pipeline.yml
Comment (mertakman) [reply eng/pipelines/update-azure-linux-pipeline.yml:98]: it is receiving                         
```yml
-build-asset-json '$(buildAssetJsonFile)'
```
at very next line

File: eng/pipelines/update-azure-linux-pipeline.yml
Comment (mertakman) [reply eng/pipelines/update-azure-linux-pipeline.yml]: fixed

File: eng/pipelines/update-azure-linux-pipeline.yml
Comment (mertakman) [reply eng/pipelines/update-azure-linux-pipeline.yml]: fixed

File: eng/pipelines/update-azure-linux-pipeline.yml
Comment (mertakman) [reply eng/pipelines/update-azure-linux-pipeline.yml]: fixed

File: eng/pipelines/update-azure-linux-pipeline.yml
Comment (mertakman) [reply eng/pipelines/update-azure-linux-pipeline.yml]: fixed

File: eng/pipelines/update-azure-linux-pipeline.yml
Comment (dagood) [reply eng/pipelines/update-azure-linux-pipeline.yml:87]: Interesting, sorry, I must have accidentally used an old tab of this PR or something--I was seeing the state of the second commit where this wasn't there yet. 😄

File: eng/pipelines/update-azure-linux-pipeline.yml
Comment (dagood) [inline eng/pipelines/update-azure-linux-pipeline.yml:22]: Nit: keep newline (including default value removal while I'm here)

```suggestion

trigger: none
```

File: eng/pipelines/update-azure-linux-pipeline.yml
Comment (dagood) [inline eng/pipelines/update-azure-linux-pipeline.yml:63]: Finishing cleaning up approval code:

```suggestion
```

File: eng/pipelines/update-azure-linux-pipeline.yml
Comment (dagood) [inline eng/pipelines/update-azure-linux-pipeline.yml:71]: This comment is about the `poll*` parameters/variables and isn't relevant here.

```suggestion
```

File: eng/pipelines/update-azure-linux-pipeline.yml
Comment (dagood) [inline eng/pipelines/update-azure-linux-pipeline.yml:5]: The new pipeline should be added to https://github.com/microsoft/go-infra/blob/main/eng/pipelines/README.md so it's easier to navigate to. Sorry I didn't spot this in the first pass.

(Line number not relevant, just wanted a resolve-able thread.)

Comment (dagood) [inline eng/pipelines/update-azure-linux-pipeline.yml:5]: Can you put a link to a test pipeline run in the PR description?

File: eng/pipelines/update-azure-linux-pipeline.yml
Comment (mertakman) [reply eng/pipelines/update-azure-linux-pipeline.yml]: fixed

File: eng/pipelines/update-azure-linux-pipeline.yml
Comment (mertakman) [reply eng/pipelines/update-azure-linux-pipeline.yml]: fixed

File: eng/pipelines/update-azure-linux-pipeline.yml
Comment (mertakman) [reply eng/pipelines/update-azure-linux-pipeline.yml]: fixed

File: eng/pipelines/update-azure-linux-pipeline.yml
Comment (mertakman) [reply eng/pipelines/update-azure-linux-pipeline.yml:5]: added

File: eng/pipelines/README.md
Comment (gdams) [inline eng/pipelines/README.md:32]:
```suggestion
* (5) [`update-azure-linux-pipeline.yml`](update-azure-linux-pipeline.yml)
```

**Review by gdams — APPROVED** [review]
LGTM - This is a good start, we'll add the pipeline triggering steps in a follow-up PR

File: eng/pipelines/README.md
Comment (dagood) [inline eng/pipelines/README.md:33]: The pipeline name is missing `-infra` in it.

File: eng/pipelines/update-azure-linux-pipeline.yml
Comment (dagood) [inline eng/pipelines/update-azure-linux-pipeline.yml:73]: In the example pipeline run https://dev.azure.com/dnceng/internal/_build/results?buildId=2525925&view=results, this step doesn't show up--has it been run?

I've cloned go-release-config-dagood to https://dev.azure.com/dnceng/internal/_library?itemType=VariableGroups&view=VariableGroupView&variableGroupId=198&path=go-release-config-mertakman for you to use for testing, if that helps. `go-release-config-mertakman` lets you set your own GitHubPAT as a secret. I would recommend making a bot GitHub account to make a simple access boundary--then, at worst, any problem is just a nuisance. Once you've saved that, passing `go-release-config-mertakman` to `goReleaseConfigVariableGroup` lets you do more in depth test runs with a custom GitHubPAT.

Adding runtime parameters is also fine for testing. (E.g. for target repo organization/name.) I don't mind leaving them in the final yml to make future testing easier too. Parameters are simpler than variables (and variable groups), so I'd recommend parameters when it's reasonable. It's not possible to make a "secret" parameter, so for GitHubPAT the variable group is necessary.

**Comment by dagood** [comment]
> we'll add the pipeline triggering steps in a follow-up PR

Before that, I think these are the more immediate next steps (which can also be done in a followup PR):

* https://github.com/microsoft/go-lab/issues/102
* https://github.com/microsoft/go-lab/issues/103

File: eng/pipelines/README.md
Comment (gdams) [reply eng/pipelines/README.md:33]: we didn't name it with that, if required we can ask the team to rename the pipeline?

Comment (dagood) [reply eng/pipelines/README.md:33]: Looks like I was able to rename it. (So the Markdown just needs to be updated.)

Comment (dagood) [reply eng/pipelines/README.md:33]:
```suggestion
  * 🚀 internal [`microsoft-go-infra-update-azure-linux`](https://dev.azure.com/dnceng/internal/_build?definitionId=1405)
```

Comment (mertakman) [reply eng/pipelines/README.md]: fixed

File: eng/pipelines/update-azure-linux-pipeline.yml
Comment (mertakman) [reply eng/pipelines/update-azure-linux-pipeline.yml:73]: The reason the step in question didn't run is that the parameter Create Pull Request on Azure Linux repo was not set to true for this pipeline run. By default, this parameter is off, and hence the associated step was skipped during execution.

Comment (mertakman) [reply eng/pipelines/update-azure-linux-pipeline.yml:91]: added.

File: eng/pipelines/update-azure-linux-pipeline.yml
Comment (mertakman) [reply eng/pipelines/update-azure-linux-pipeline.yml:35]: fixed

File: eng/pipelines/README.md
Comment (mertakman) [reply eng/pipelines/README.md]: fixed

**Review by dagood — APPROVED** [review]
We did some work on this in a call, and got this successful test build: https://dev.azure.com/dnceng/internal/_build/results?buildId=2531077&view=results
Which made this PR: https://github.com/mertakman/azurelinux/pull/14
(This branch doesn't have https://github.com/microsoft/go-infra/pull/161, and the output looks good other than that.)

**Pull request merged**
